### PR TITLE
Add information on lastOptSubIndex

### PR DESCRIPTION
**Pull request by a7ehuo**
This update captures a discussion on Slack on how to use `lastOptIndex` and `lastOptSubIndex` to control `performTransformation` to skip optional opt passes to narrow down compilation issues.

Co-Authored-By: Devin Papineau <devin@ajdmp.ca>

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

### CODE REVIEW

**Comment by a7ehuo** [comment]
@0xdaryl May I ask you to review? Thank you!
@jdmpapin fyi

**Comment by a7ehuo** [comment]
@0xdaryl @jdmpapin Just following up. Is there any review comment on this update? Thanks

File: doc/compiler/ProblemDetermination.md [hunk 1]
```diff
@@ -216,13 +216,45 @@ compilation altogether, using the `exclude=method` parameter:
...
+By adjusting `N` in `lastOptIndex=N` that tells the compiler to perform `N` optional
+transformations and observing when the test passes and fails, it is possible to narrow
+down the failing transformation using a binary-search approach. Once the issue is
+narrowed down to one opt pass, then try searching for the bad transformation with
+`lastOptSubIndex`.
+
+For example, if the issue is narrowed down to `lastOptIndex=28`, `optSubIndex` will
+be printed out in the [compilation log](optimizer/IntroReadLogFile.md). Next `lastOptIndex=28`
+and `lastOptSubIndex=3` can be set together to continue to binary search the sub opt
+within the bad transformation.
```
Comment (hzongaro) [inline doc/compiler/ProblemDetermination.md:241]: I would suggest rewording "to binary search the sub opt within the bad transformation" to "the binary search for the bad transformation within the last optimization pass."

File: doc/compiler/ProblemDetermination.md [hunk 2]
```diff
@@ -216,13 +216,45 @@ compilation altogether, using the `exclude=method` parameter:
...
+compilations to avoid overhead in the overwhelmingly common non-tracing case.
+
+`performTransformation` is [controlled by comparing](https://github.com/eclipse/omr/blob/f5a790b68037be3a48b6e5c4f7e8ae6ad3e30f00/compiler/ras/Debug.cpp#L477)
+the current `optIndex` with `firstOptIndex` and  `lastOptIndex` options.
+
+By adjusting `N` in `lastOptIndex=N` that tells the compiler to perform `N` optional
+transformations and observing when the test passes and fails, it is possible to narrow
+down the failing transformation using a binary-search approach. Once the issue is
+narrowed down to one opt pass, then try searching for the bad transformation with
+`lastOptSubIndex`.
```
Comment (hzongaro) [inline doc/compiler/ProblemDetermination.md:236]: You might also want to mention that `lastOptSubIndex` also controls `performTransformation` at this location https://github.com/eclipse/omr/blob/f5a790b68037be3a48b6e5c4f7e8ae6ad3e30f00/compiler/ras/Debug.cpp#L523-L525

File: doc/compiler/ProblemDetermination.md [hunk 3]
```diff
@@ -216,13 +216,45 @@ compilation altogether, using the `exclude=method` parameter:
...
+[     5]  28.1    O^O VALUE PROPAGATION: Setting element width for array [00007F2D3F004B70] to 0
+[     6]  28.2    O^O VALUE PROPAGATION: Removing redundant null check node [00007F2D3F004DF0]
+[     7]  28.3    O^O VALUE PROPAGATION: Setting element width for array [00007F2D3F004B70] to 0
+[     8]  28.4    O^O VALUE PROPAGATION: Replace PassThrough node [00007F2D3F004DA0] with its child in its parent [00007F2D3F004DF0]
+[     9]  28.5    O^O VALUE PROPAGATION: Replacing n26n acall of <jitLoadFlattenableArrayElement>
+```
+
+`lastOptIndex=-1` is an even stronger property than `optLevel=noOpt`.
+
+Since `lastOptTransformationIndex` uses a single static counter for all compilations, it is
```
[the remark this answers is not in the document]
Comment (a7ehuo) [reply doc/compiler/ProblemDetermination.md:253]: I could update the paragraph as below. I think even it's not often used (I haven't used it either), it might be good to point out here what it is and why it's not used often so that people don't have to try it if it's not particular useful. 

```
Another option related to the index of an optimization transformation is `lastOptTransformationIndex`,
which is the index of the last optimization transformation to perform.
However it uses a single static counter for all compilations. Therefore it is
not very reliable or often used. It can really only be used in situations where the JIT will not run any
compilations concurrently, and where it will run the same sequence of compilations every time.
(That's on top of the possibility that some non-determinism might cause a particular
compilation to do a different sequence of transformations or even of opt passes.)
```

File: doc/compiler/ProblemDetermination.md [hunk 4]
```diff
@@ -216,13 +216,45 @@ compilation altogether, using the `exclude=method` parameter:
...
-check called `performTransformation` that can be controlled by the option
-`lastOptIndex=N`. This tells the compiler to perform `N` optional transformations,
-then proceed to only do required operations. In this sense, `lastOptIndex=-1` is an
-even stronger property than `optLevel=noOpt`.
+check called [performTransformation](https://github.com/eclipse/omr/blob/f5a790b68037be3a48b6e5c4f7e8ae6ad3e30f00/compiler/compile/OMRCompilation.hpp#L271-L274)
+([more details](https://github.com/eclipse/omr/blob/8f5212d52b533cdf9d6de9c2bd8a3e01e83e8b9d/compiler/compile/OMRCompilation.hpp#L170-L226)).
 
-By binary searching `lastOptIndex=` you can find the optimization that failed.
+`performTransformation` is [enabled by option](https://github.com/eclipse/omr/blob/8f5212d52b533cdf9d6de9c2bd8a3e01e83e8b9d/compiler/compile/OMRCompilation.hpp#L242)
+`TR_TraceOptDetails` or `TR_CountOptTransformations`.
```
[the remark this answers is not in the document]
Comment (hzongaro) [reply doc/compiler/ProblemDetermination.md:223]: Yes.  In addition `TR_TraceOptDetails` is also set if the `optDetails` option is specified, and `TR_CountOptTransformations` is enabled if `lastOptSubIndex` is specified.  I hope that covers all the cases - today, at least!

File: doc/compiler/ProblemDetermination.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (hzongaro) [reply doc/compiler/ProblemDetermination.md:253]: I think that looks good, Annabelle.

**Comment by a7ehuo** [comment]
@hzongaro All comments are addressed. Ready for another review!

File: doc/compiler/ProblemDetermination.md [hunk 5]
```diff
@@ -216,13 +216,54 @@ compilation altogether, using the `exclude=method` parameter:
 ### Identifying the failing Optimization.
 
 Most transformations that can be elided are guarded in the source code by a special
-check called `performTransformation` that can be controlled by the option
-`lastOptIndex=N`. This tells the compiler to perform `N` optional transformations,
-then proceed to only do required operations. In this sense, `lastOptIndex=-1` is an
-even stronger property than `optLevel=noOpt`.
+check called [performTransformation](https://github.com/eclipse/omr/blob/f5a790b68037be3a48b6e5c4f7e8ae6ad3e30f00/compiler/compile/OMRCompilation.hpp#L271-L274)
+([more details](https://github.com/eclipse/omr/blob/8f5212d52b533cdf9d6de9c2bd8a3e01e83e8b9d/compiler/compile/OMRCompilation.hpp#L170-L226)).
```
Comment (hzongaro) [inline doc/compiler/ProblemDetermination.md:220]: It occurs to me that this doesn't mention the fact that [`performOptimization` might entirely skip on optimization based on the settings of `firstOptIndex` and `lastOptIndex`](https://github.com/eclipse/omr/blob/8f5212d52b533cdf9d6de9c2bd8a3e01e83e8b9d/compiler/optimizer/OMROptimizer.cpp#L1670-L1671) if it's not a required optimization.  In such cases, we'll never even get to the checks in `performTransformation`.

File: doc/compiler/ProblemDetermination.md [hunk 6]
```diff
@@ -216,13 +216,54 @@ compilation altogether, using the `exclude=method` parameter:
...
+the current `optIndex` with `firstOptIndex` and  `lastOptIndex` options.
+
+`lastOptIndex` - the index of the last optimization - can be used to narrow things
+down to a particular optimization pass. By adjusting `N` in `lastOptIndex=N` that tells
+the compiler to perform `N` transformation passes and observing when the test passes
+or fails, it is possible to narrow down the failing transformation using a binary-search
+approach. Once the issue is narrowed down to one opt pass, then try searching for the
+bad transformation with `lastOptSubIndex`. `lastOptSubIndex` is used to narrow things
+down to a particular optional transformation within the last optimization. `lastOptSubIndex`
+also controls `performTransformation` as mentioned on the above.
```
Comment (hzongaro) [inline doc/compiler/ProblemDetermination.md:242]: Typo - should read "as mentioned above"

**Review by hzongaro — COMMENTED** [review]
Sorry for this late additional comment about `performOptimization`.  It hadn't occurred to me before that the existing documentation only talked about `performTransformation`, but not the higher level `performOptimization`.

File: doc/compiler/ProblemDetermination.md (hunk 6 above)
Comment (a7ehuo) [reply doc/compiler/ProblemDetermination.md]: Addressed in 75e009b

File: doc/compiler/ProblemDetermination.md (hunk 5 above)
Comment (a7ehuo) [reply doc/compiler/ProblemDetermination.md:220]: Addressed in 75e009b

**Comment by a7ehuo** [comment]
@hzongaro All comments are addressed. Ready for another review. Thanks!

**Review by hzongaro — APPROVED** [review]
Looks good.  Thanks!

**Review by 0xdaryl — APPROVED** [review]
